### PR TITLE
metal: remove two per-step decode syncs on the spec path

### PR DESCRIPTION
**Pull request by ehartford**
Drain-peeling on the Qwen3.8 Metal decode loop, found by cProfile over the seeded offline bench and verified byte-identical at each step.

1. **`mamba_hybrid.postprocess_state`**: boolean-mask indexing (`idx_mapping[valid]`) lowers to `nonzero()` on MPS — a full device-queue drain measured at **42 ms/step, half the decode wall**. Replaced with the sentinel-redirect scatter already used by `input_batch.post_update` (int-index gather + where + scatter, sync-free; invalid rows rewrite slot 0's own value).
2. **`metal_attn` metadata build**: `seq_lens_cpu = m.seq_lens.to("cpu")` ran per build (~24/step) and the first one absorbed the entire previous GPU step (~40 ms). Under serial scheduling the scheduler has already consumed the prior step's acceptance, so `seq_lens_cpu_upper_bound` (committed + scheduled) is exact — build uses it, keeping the synced pull only under async scheduling where the bound can overshoot into attending stale KV rows.

Qwen3.8 seeded bench (shipped sampling defaults): essay 24.4 → 25.7 tok/s, GSM8K 39.4 → 41.0, **outputs byte-identical** through both changes. The remaining per-step wait is a single output-consumption event sync that serial scheduling cannot avoid; the notebook entry maps the remaining CPU-encode and GPU-kernel levers.

42 kernel/worker tests pass; ruff clean.

## Summary by CodeRabbit

* **Performance Improvements**
  * Reduced unnecessary synchronization during Metal attention, state processing, metadata construction, and positional preparation.
  * Added support for larger 256-dimension attention workloads and strided KV-cache layouts.
  * Improved handling of hybrid attention and speculative decoding metadata.
* **Reliability**
  * Made state updates deterministic when entries share indices.
  * Preserved correct behavior across synchronous and asynchronous scheduling paths.
* **Compatibility**
  * Improved Qwen vision-model processor compatibility and reduced unnecessary vision dependencies in text-only environments.
* **Documentation**
  * Expanded optimization notes with benchmarks, validation results, and follow-up work.

### CODE REVIEW

File: vllm/v1/worker/gpu/model_states/mamba_hybrid.py [hunk 1]
```diff
@@ -307,13 +307,21 @@ def postprocess_state(
...
+                # own gathered value (a no-op; duplicates write the same
+                # value), the same pattern as input_batch.post_update.
+                idx = idx_mapping.to(torch.long)
+                valid = idx >= 0
+                safe = torch.where(valid, idx, torch.zeros_like(idx))
+                vals = torch.clamp(num_sampled[:n], min=1).to(
+                    self.num_accepted_tokens_gpu.dtype
                 )
+                cur = self.num_accepted_tokens_gpu[safe]
+                self.num_accepted_tokens_gpu[safe] = torch.where(valid, vals, cur)
```
[the remark this answers is not in the document]
Comment (ehartford) [reply vllm/v1/worker/gpu/model_states/mamba_hybrid.py]: Confirmed and fixed in 74fecbbde. The update is now additive: sentinel rows contribute a zero delta and valid rows contribute (vals - cur), applied with integer scatter_add — deterministic under duplicate indices, so a sentinel colliding with a genuine request state 0 can no longer win the write race. Semantics verified on-box with a mixed sentinel/valid-0 batch ([7,9,11,13] -> [4,9,2,13]); the seeded bench is unchanged with byte-identical outputs. Still sync-free (no nonzero on the device path).
